### PR TITLE
Xz/plugin selection

### PR DESCRIPTION
**Pull request by zhangxu0307**
1. add auto plugin selection
2. add embedding module
3. modify plugin registry
4. revise some ut

### CODE REVIEW

File: taskweaver/utils/embedding.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+            try:
+                from sentence_transformers import SentenceTransformer
+
+                self.embedding_model = SentenceTransformer(self.config.embedding_model)
+            except Exception:
+                raise Exception(
+                    "Package sentence_transformers is required for using embedding. "
+                    "Please install it using pip install sentence_transformers",
+                )
+        elif self.config.embedding_model_type == "openai":
```
Comment (Jack-Q) [inline taskweaver/utils/embedding.py:52]: Seems the initialization of this part is rather overlapped with the LLM module, not sure whether we can somehow have these two part merged into one place? For example, the AzureAD is not supported in this part.

**Comment by Jack-Q** [comment]
Besides, it seems that planner, code executor are all require plugin registry to load the plugin within each session, do we have any approach to sync and notify the plugin selection in different modules?

**Comment by zhangxu0307** [comment]
> Besides, it seems that planner, code executor are all require plugin registry to load the plugin within each session, do we have any approach to sync and notify the plugin selection in different modules?

Following our discussion, we have decided to enable automatic plugin selection exclusively for the code generator due to the following reasons:
1. In the case of the Planner, it necessitates complete information on all plugins for effective task planning.
2. As for the Code Executor, loading all plugins might not cause any side effects, as the code generator is only aware of the selected plugins.

File: taskweaver/plugin/plugin_selection.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+        user_query_embedding = np.array(self.embedding_generator.get_embedding(user_query))
+
+        similarities = []
+
+        assert (
+            len(
+                self.plugin_registry.get_list(),
+            )
+            >= top_k
+        ), "the number of plugins to be selected should be larger than top_k"
```
[the remark this answers is not in the document]
Comment (zhangxu0307) [reply taskweaver/plugin/plugin_selection.py]: solved in # 4622e30

File: taskweaver/utils/embedding.py [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+                self.client = AzureOpenAI(
+                    api_version=self.llm_config.api_version,
+                    azure_endpoint=self.llm_config.api_base,
+                    api_key=self.llm_config.api_key,
+                )
+            elif api_type == "openai":
+                self.client = OpenAI(
+                    base_url=self.llm_config.api_base,
+                    api_key=self.llm_config.api_key,
+                )
```
[the remark this answers is not in the document]
Comment (zhangxu0307) [reply taskweaver/utils/embedding.py]: Try to unifiy with llm_api in # https://github.com/microsoft/TaskWeaver/commit/3f796df15277fe376b223e7ccc1c2bf79072080a

**Comment by 3e4shi** [comment]
3f796df15277fe376b223e7ccc1c2bf79072080a

**Review by Jack-Q — APPROVED** [review]

**Pull request merged**